crypto/rand/rand_win.c

```diff
@@ -62,20 +62,23 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
 # endif
 
 # ifdef USE_BCRYPTGENRANDOM
-    bytes_needed = rand_pool_bytes_needed(pool, 1 /*entropy_factor*/);
-    buffer = rand_pool_add_begin(pool, bytes_needed);
-    if (buffer != NULL) {
-        size_t bytes = 0;
-        if (BCryptGenRandom(NULL, buffer, bytes_needed,
+    static HMODULE bcrypt = ::GetModuleHandleA("bcrypt.dll");
+    if (bcrypt != NULL) {
+        bytes_needed = rand_pool_bytes_needed(pool, 1 /*entropy_factor*/);
+        buffer = rand_pool_add_begin(pool, bytes_needed);
+        if (buffer != NULL) {
+            size_t bytes = 0;
+            if (BCryptGenRandom(NULL, buffer, bytes_needed,
                             BCRYPT_USE_SYSTEM_PREFERRED_RNG) == STATUS_SUCCESS)
-            bytes = bytes_needed;
+                bytes = bytes_needed;
 
-        rand_pool_add_end(pool, bytes, 8 * bytes);
-        entropy_available = rand_pool_entropy_available(pool);
+            rand_pool_add_end(pool, bytes, 8 * bytes);
+            entropy_available = rand_pool_entropy_available(pool);
+        }
+        if (entropy_available > 0)
+            return entropy_available;
     }
-    if (entropy_available > 0)
-        return entropy_available;
-# else
+# endif
```

Comment:
If `BCryptGenRandom` is used, then `CryptGenRandom` shouldn't be.  Why use the legacy call if the later one exists?  Underneath they are the same.

Do you have an example where the former exists but fails and the latter doesn't?

Reply:
> If BCryptGenRandom is used, then

The change is not about that. It's about: 

> check if bcrypt.dll is present before using BCryptGenRandom

that is, `BCryptGenRandom` shouldn't be called if bcrypt.dll isn't present (on WinXP). This change doesn't try to do LoadLibrary+GetProcAddress, but rather expects that final application uses delay loading for bcrypt at link time. This way final executable won't be restricted to Vista+ builds if it links to openssl.